engine/common/splitter/network/network.go

```diff
@@ -0,0 +1,96 @@
+package network
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	splitterEngine "github.com/onflow/flow-go/engine/common/splitter"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/network"
+	"github.com/rs/zerolog"
+)
+
+type Network struct {
+	mu        sync.RWMutex
+	net       module.Network
+	log       zerolog.Logger
+	splitters map[network.Channel]*splitterEngine.Engine // stores splitters for each channel
+	conduits  map[network.Channel]network.Conduit        // stores conduits for all registered channels
+	subMngr   network.SubscriptionManager
+}
+
+func NewNetwork(
+	net module.Network,
+	log zerolog.Logger,
+	subMngr network.SubscriptionManager,
+) (*Network, error) {
+	e := &Network{
+		net:       net,
+		splitters: make(map[network.Channel]*splitterEngine.Engine),
+		conduits:  make(map[network.Channel]network.Conduit),
+		log:       log,
+		subMngr:   subMngr,
+	}
+
+	return e, nil
+}
+
+// Register will subscribe the given engine with the spitter on the given channel, and all registered
+// engines will be notified with incoming messages on the channel.
+// The returned Conduit can be used to send messages to engines on other nodes subscribed to the same channel
+func (n *Network) Register(channel network.Channel, e network.Engine) (network.Conduit, error) {
```

Comment:
Shouldn't the splitter network should automatically register a special `RelayEngine` every time a call for Register is made by any other engine. This way, the `RelayEngine` would be registered for all channels used by the engines of the access node without having to explicitly know beforehand what those are.

Reply:
No, the splitter engine conceptually has nothing to do the `RelayEngine` at all. What the splitter engine does is it forwards every message it receives to every engine that has registered with it.

The splitter engine indeed enables the use case that we need for the unstaked access node, which involves registering a relay engine alongside each engine for all the different channels, but in theory at least there may be other potential usecases where a splitter engine may be useful. Namely, any use case where more than one engine might want to receive messages on the same channel. This is why i have defined the engine in the `common` folder rather than `access`.

Reply:
From the splitter engine's perspective, the relay engine is just one of many engines that will register with it.

Reply:
This would mean the `RelayEngine` would have to `poll` for all the channels and the scaffold has to make sure that the `RelayEngine` is the last of the engine that is started else it won't get the whole channel list.
A simpler solution would be to just add the RelayEngine as another engine each time any engine subscribes.

Reply:
Imo, I think the better solution is for the setup in the access node `main.go` to ensure that relay engine is created last, exactly you say. After all, the job of the `main.go` is to create all the engines and hook them up properly. This would not be the first example where the instantiation of an engine depends on the existence of another engine (ie one engine is passed into the constructor of another).

Furthermore, the list of channels should be directly passed into the relay engine constructor, and we can get this channel list via the `Channels()` method on subscription manager.

Reply:
I really want to avoid coupling the splitter engine with the relay engine. I think the splitter engine has the potential to be a much more general purpose component

Reply:
imho, we have a use case for a relay engine that should be automatically tagged for all channels as and when they are being subscribed and unsubscribed (there is an unregister but it happens rarely). We don't really have a use cast to have arbitrary number of engines subscribing for the same channel. Hence, splitter seems a bit of an overkill to me and also introduces implicit assumptions for the relay engine. BUT, I don't feel very strong about it one way or the other and good with this design too.

Reply:
I understand what you're saying, but its still my opinion that we should try to separate the two engines. Because otherwise, why bother implementing two separate engines at all? We could just take all of the logic from the relay engine and just put it into the splitter engine.

I also want to point out that the assumptions for the relay engine are not really implicit, because the list of channels will be passed in to its constructor. Therefore, it will know precisely the list of channels it needs to relay messages on. So we could even create the relay engine before we create any of the other engines, as long as we know which channels we want to relay messages on.

Reply:
OK, this may be a comment sitting across this PR and #984 but:

I see the appeal in a generic splitter, but I would also appreciate a strong godoc paragraph at the beginning of the engine that lays out where and when the splitter receives its channel processing obligations over time, and how it maintains them.

The topology of the splitter is clear, but to Vishal's point the channel responsibilities over the lifecycle of the component are a bit harder to tease out when channels are per-method parameters.

Another way to do this would be to write a mixed splitter / relay [runnable example](https://blog.golang.org/examples) that would demonstrate this over time (esp. w/ unsubscribes)

Reply:
https://github.com/onflow/flow-go/issues/992